### DISCUSSION TITLE
17 gennaio 2024 B - esercizio 1

### DISCUSSION THREAD

**Discussion by IonutCicio** [sapienzastudentsnetwork/basi-di-dati-1#92] [Esami]
![image](https://github.com/sapienzastudentsnetwork/bdi2324/assets/40034429/10da4913-a23f-4437-adb9-c4005779606c)

**Comment by Frisbon** [comment]
1a)
![image](https://github.com/sapienzastudentsnetwork/bdi2324/assets/84731341/ac1d9d7a-07cc-4041-9b34-3e90ef28ee41)
1b)
Dato che mi serve il personale **SOLO** da Ancona, devo eliminare quelli che sono partiti da Ancona e da altri posti. L’idea é di togliere dall’equipaggio, coloro che sono partiti da posti ≠ da Ancona, in questo modo, Bob che é partito da Roma e da Ancona verrá cancellato mentre Rob che é partito solo da Ancona no.

![image](https://github.com/sapienzastudentsnetwork/bdi2324/assets/84731341/11bf18de-2c32-4165-9f58-4e60c00a93d9)

Grazie @CuriousCI per la correzione, (ora é giusta?)

**Reply by IonutCicio** [reply]
Non puoi fare $\sigma_{\text{Arrivo = Pescara}}$ e  $\sigma_{\text{Partenza } \neq \text{ Pescara}}$ perché $\text{TRATTA.Partenza}$ e $\text{TRATTA.Arrivo}$ si riferiscono a $\text{Sigle}$ di porti.

Quindi dovresti prima fare $\rho_{\text{Ancona } \leftarrow \text{ Sigla}}( \pi_{\text{Sigla}}( \sigma_{\text{Città = Ancona}}(\text{PORTO})) )$ _(in maniera simile per la seconda query)_ 

> [!NOTE]
> Per come è costruito lo schema logico potrebbe succedere che una città ha più di un porto

> [!TIP]
> Ci sono altri esercizi simili, come quello degli aeroporti, in cui è abbastanza comune fare questo errore, fateci attenzione!

**Comment by gioele-m-l** [comment]
>Se trovate incorrettezze fatemi sapere thx <3.

1a)
1. Ricavo le _sigle_ dei ***Porti*** con _città_="Pescara":

$$PortiPescara = \pi_{Sigla}(\sigma_{Città="Pescara"}(Porto))$$

2. Ora trovo le _sigle_ delle ***Tratte*** dove l'_arrivo_ coincide con una delle _sigle_ dei ***PortiPescara***, eseguo anche una selezione sull'_orario_:

$$TrattaPPescO = \pi_{Tratta.Sigla}(\sigma_{Orario>12:00}(Tratta\bowtie_{Arrivo=PortiPescara.Sigla}PortiPescara))$$

3. Infine, per trovare le _date_ dei ***Viaggi*** di interesse (quelli con arrivo a Pescara, partenza dopo le 12:00, di viaggi tra il 2018 e il 2019), faccio prima un theta join tra le tabelle ***Viaggio*** e ***TrattaPPescO***, sul risultato di esso applico una selezione sulle date e per finire proietto gli attributi _Viaggio.Data_ delle tuple ottenute:

$$Result = \pi_{Data}(\sigma_{Data>=01/01/2018 \land Data<=31/12/2019}(Viaggio \bowtie_{Viaggio.SiglaTratta=TrattaPPescO.Sigla}TrattaPPescO))$$


1b)
1. Per prima cosa ottengo le _sigle_ dei ***Porti*** non situati ad "Ancona":

$$NoPortiAncona = \pi_{Sigla}(\sigma_{Città\neq"Ancona"}(Porto))$$

2. Ora, tramite theta join con ***Tratta***, ottengo le **tratte** la cui _partenza_ corrisponde alla _sigla_ di uno dei **porti** di ***NoPortiAncona***:

$$TratteNoPartAncona = \pi_{Tratta.Sigla}(Tratta\bowtie_{Partenza=NoPortiAncona.Sigla}NoPortiAncona)$$

3. Faccio sostanzialmente la cosa di prima però ottenendo gli _ID_ dei ***Viaggi*** che non partono da "Ancona", tramite le **tratte** che non partono da "Ancona":

$$ViaggiNoPartAncona = \pi_{Viaggio.ID}(Viaggio\bowtie_{Viaggio.SiglaTratta=TratteNoPartAncona.Sigla}TratteNoPartAncona)$$

4. Ora calcolo il ***Personale*** che ha viaggiato in uno dei **viaggi** con _partenza_ diversa da "Ancona":

$$PersPartNoAncona = \pi_{Personale.CF}(Personale\bowtie_{Personale.ID=Equipaggio.Pers}Equipaggio\bowtie_{Equipaggio.Viaggio=ViaggiNoPartAncona.ID}ViaggiNoPartAncona)$$

5. Infine per calcolare il risultato finale (dati del ***Personale*** di ***Viaggi*** con _partenze_ solo da "Ancona") faccio la differenza tra la tabella ***Personale*** e la tabella ***PersPartNoAncona*** (opportunamente sistemata):

$$Result = Personale - (PersPartNoAncona\bowtie Personale)$$
